### ISSUE TITLE
hdf5: support thread-safety and c++ bindings

### ISSUE THREAD

**Issue by slayoo** [Homebrew/homebrew-science#1888]
Currently in hdf5.rb the thread-safety option is assumed to be mutually exclusive with the cxx option.
This is linked with the fact that the configure script suggest that this combination is not supported, although it also says that one can compile it by using the --enable-unsupported option.

According to the following explenation from HDF5 developers:

http://hdf-forum.184993.n3.nabble.com/hdf-forum-HDF5-enable-cxx-enable-threadsafe-conflict-ubuntu-debian-packages-td194242.html

the reason why these two options are considered unsupported is not to cause confusion - the thread-safety relates only the C API.

Perhaps it is then better to only issue a warning suggesting to the users not to interpret "--enable-threadsafe --enable-cxx" as meaning that the thread-safety is supported under C++.

In Debian, they do use --enable-unsupported --enable-cxx --enable-threadsafe:
https://sources.debian.net/src/hdf5/1.8.13%2Bdocs-15/debian/rules/#L95

To sum up - please do support both thread-safety and cxx!
Thanks,
Sylwester

**Comment by slayoo** [comment]
Other option would be to simply mimic hdf5 configure's behaviour - i.e. to add an -enable-unsupported option for the brew's hdf5 script.

**Comment by iml** [comment]
Sorry for the delay. This is useful information, and we could probably make this change in addition to enabling cxx by default, since it's a common conditional dependency for many formulae.

Would `--with-flavor=serial` is also be required, and does this conflict with the other `openmpi` and `mpich` flavors? Are any of these mutually exclusive?

**Comment by iml** [comment]
I can't see any errors or test failures building with both parallel and thread-safety, though I would expect either one of those to impact the other. Perhaps a simple warning would be more appropriate. It seems like a different rationale than not supporting enable-cxx.

**Comment by iml** [comment]
OK, according to [this thread](http://hdf-forum.184993.n3.nabble.com/HDF5-parallel-and-threadsafe-td1701166.html) the rationale is actually somewhat similar and the two might actually work together just fine.

**Comment by jakirkham** [comment]
Found a response from an HDF5 dev a while back that seems much clearer IMHO than the other to. Figured I'd include it here if people come looking for why this would be ok.

ref: https://lists.hdfgroup.org/pipermail/hdf-forum_lists.hdfgroup.org/2009-June/000135.html
